common/set_benchmark.cpp

```diff
@@ -0,0 +1,366 @@
+    for (ssize_t i = 0; i < lookup_keys_size;) {
+      benchmark::DoNotOptimize(i);
+
+      s.BenchErase(lookup_keys[i]);
+      benchmark::ClobberMemory();
+
+      bool inserted = s.BenchInsert(lookup_keys[i]);
+      CARBON_DCHECK(inserted);
+      i += static_cast<ssize_t>(inserted);
+    }
+  }
+}
+MAP_BENCHMARK_ONE_OP(BM_SetEraseInsertHitPtr, HitArgs);
+
+// NOLINTBEGIN(bugprone-macro-parentheses): Parentheses are incorrect here.
+#define MAP_BENCHMARK_OP_SEQ_SIZE(NAME, KT)                  \
+  BENCHMARK(NAME<Set<KT>>)->Apply(SizeArgs);                 \
+  BENCHMARK(NAME<absl::flat_hash_set<KT>>)->Apply(SizeArgs); \
+  BENCHMARK(NAME<llvm::DenseSet<KT>>)->Apply(SizeArgs);      \
+  BENCHMARK(NAME<llvm::DenseSet<KT, CarbonHashDI<KT>>>)->Apply(SizeArgs)
+// NOLINTEND(bugprone-macro-parentheses)
+
+#define MAP_BENCHMARK_OP_SEQ(NAME)       \
+  MAP_BENCHMARK_OP_SEQ_SIZE(NAME, int);  \
+  MAP_BENCHMARK_OP_SEQ_SIZE(NAME, int*); \
+  MAP_BENCHMARK_OP_SEQ_SIZE(NAME, llvm::StringRef)
+
+// This is an interesting, somewhat specialized benchmark that measures the cost
+// of inserting a sequence of keys into a set up to some size and then inserting
+// a colliding key and throwing away the set.
+//
+// This is an especially important usage pattern for sets as a large number of
+// algorithms essentially look like this, such as collision detection, cycle
+// detection, de-duplication, etc.
+//
+// It also covers both the insert-into-an-empty-slot code path that isn't
+// covered elsewhere, and the code path for growing a table to a larger size.
+//
+// This is the second most important aspect of expected set usage after testing
+// for presence. It also nicely lends itself to a single benchmark that covers
+// the total cost of this usage pattern.
+//
+// Because this benchmark operates on whole sets, we also compute the number of
+// probed keys for Carbon's set as that is both a general reflection of the
+// efficacy of the underlying hash function, and a direct factor that drives the
+// cost of these operations.
+template <typename SetT>
+static void BM_SetInsertSeq(benchmark::State& state) {
+  using SetWrapperT = SetWrapper<SetT>;
+  using KT = typename SetWrapperT::KeyT;
+  constexpr ssize_t LookupKeysSize = 1 << 8;
+  auto [keys, lookup_keys] =
+      GetKeysAndHitKeys<KT>(state.range(0), LookupKeysSize);
+
+  // Now build a large shuffled set of keys (with duplicates) we'll use at the
+  // end.
+  ssize_t i = 0;
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(i);
+
+    SetWrapperT s;
+    for (auto k : keys) {
+      bool inserted = s.BenchInsert(k);
+      CARBON_DCHECK(inserted) << "Must be a successful insert!";
+    }
+
+    // Now insert a final random repeated key.
+    bool inserted = s.BenchInsert(lookup_keys[i]);
+    CARBON_DCHECK(!inserted) << "Must already be in the map!";
+
+    // Rotate through the shuffled keys.
+    i = (i + static_cast<ssize_t>(inserted)) & (LookupKeysSize - 1);
```

Comment:
Isn't `inserted` false in this case? 
```suggestion
    i = (i + static_cast<ssize_t>(!inserted)) & (LookupKeysSize - 1);
```